### PR TITLE
Feature/adaptive workers

### PR DESCRIPTION
**Pull request by shuds13**
Draft PR for variable resource workers.

For code comparison.

-----------------------------------------------------------

### Major restructuring for adaptive worker resources.

Resources are divided into **_resource sets_**. By default, the no. of workers.
One resource set can include nodes and partitions on nodes (slots).
To use variable resources, a `resource_sets` field is added to `gen_specs['out']`

The generator function provides an integer value for `resource_sets`.
Allocation function provisions sufficient worker resource (manager assigns on send)
Resources is separated from executor but linked to it. Configured by `libE_specs`.
Resources provides nodes and partitions (slots) on the node to each worker.

Regression tests using:
 - test_uniform_sampling_with_different_resources.py
 - test_persistent_sampling_CUDA_variable_resources.py

Currently alloc functions using:
 - give_sim_work_first.py
 - start_only_persistent.py

Note also:
 - `alloc_support` is now a class.
 - Usage in alloc functions is optional - depends on whether `resource_sets` field is used.
 - Resources can be turned off altogether.

Todo:
 - Restructure resources (see branch feature/restructure_resources)
 - Restructure alloc functions (will look [something like](https://github.com/Libensemble/libensemble/blob/feature/adaptive_workers_alloc_restructure/libensemble/alloc_funcs/start_only_persistent_using_wrapper.py))

Structure discussion required:
 - Is separation of resources from Executor correct decision?
 - Using class variables for executor/resources - or pass via libE_info (inc. to allocation function).
 - Resources structure (inheritence / composition)

Terminology / API Discussion required.
 - Terminology for resource info that can be extracted by user (`slots_on_node` etc...)
 - Terminology - custom_info/resource_info? -- For overriding auto-detection (e.g. num_nodes cores_per_node etc...) 
 - Also resource options like central_mode, zero_resource_workers - should they be sub-dictionary - or start with a prefix.... e.g., resource_*
 - The term 'resource sets'
 - Maybe InsufficientResourcesException not clear that it means current available resources so need to wait for some to free up.  eg e.g InsufficientFreeResources or InsufficientAvailResources or InsufficientCurrentResources

### CODE REVIEW

File: libensemble/alloc_funcs/start_persistent_consensus.py [hunk 1]
```diff
@@ -152,14 +156,14 @@ def start_consensus_persistent_gens(W, H, sim_specs, gen_specs, alloc_specs, per
...
+                persis_info[wid].update({'curr_H_ids': [l_H_id, r_H_id]})
 
     # If all gens at consensus, distribute data to all gens
-    if num_gens_at_consensus == alloc_specs['user']['num_gens']:
+    if num_gens_at_consensus == user['num_gens']:
 
         assert num_gens_at_consensus == len(avail_persis_worker_ids), \
             'All gens must be available, only {}/{} are though...'.format(
-                len(avail_worker_ids), len(num_gens_at_consensus))
+                len(avail_persis_worker_ids), len(num_gens_at_consensus))
```
Comment (jmlarson1) [inline libensemble/alloc_funcs/start_persistent_consensus.py:166]: Good catch!

File: docs/data_structures/libE_specs.rst [hunk 2]
```diff
@@ -85,6 +85,73 @@ Specifications for libEnsemble::
...
+        'disable_resource_manager' [boolean]:
+            Disable the built-in resource manager. If this is True, then resource detection
+            and/or assignment of resources to workers is disabled. This also means that
+            any entries in the ``resource_info`` option will be ignored.
+            Default: False
+        'num_resource_sets' [int]:
+            The total number of resource sets. Resources will be divided into this number.
+            Default: None. If None, resources will be divided by workers (excluding
+            zero_resource_workers).
+        'enforce_worker_core_bounds' [boolean]:
```
Comment (jlnav) [inline docs/data_structures/libE_specs.rst:97]: Should probably mention Default: False

File: libensemble/manager.py [hunk 3]
```diff
@@ -398,6 +433,7 @@ def _update_state_on_worker_msg(self, persis_info, D_recv, w):
             if w in self.persis_pending:
                 self.persis_pending.remove(w)
                 self.W[w-1]['active'] = 0
+            self._freeup_resources(w)  # SH TODO - check persistent pending does not screw this up
```
Comment (jlnav) [inline libensemble/manager.py:436]: Is this being checked somewhere, or are we not worrying about it for now?

File: libensemble/resources/mpi_resources.py [hunk 4]
```diff
@@ -4,158 +4,243 @@
...
+        return var
+
+
+def get_MPI_variant():
+    """Returns MPI base implementation
+
+    Returns
+    -------
+    mpi_variant: string:
+        MPI variant 'aprun' or 'jsrun' or 'mpich' or 'openmpi'
```
Comment (jlnav) [inline libensemble/resources/mpi_resources.py:39]: or 'srun'

File: libensemble/resources/scheduler.py [hunk 5]
```diff
@@ -0,0 +1,238 @@
...
+
+        If the resources required are more than one node, then will
+        attempt to find an even split. If no even split is possible,
+        then enough full nodes will be allocated to cover the requirement.
+
+        Returns a list of resource sets ids. A return of None implies
+        insufficient resources.
+        """
+
+        # SH TODO: Remove commented print statements
```
Comment (jlnav) [inline libensemble/resources/scheduler.py:65]: TODO reminder?

File: docs/FAQ.rst [hunk 6]
```diff
@@ -15,23 +15,48 @@ running with multiprocessing and multiple workers specified.**
...
 make sure that ``libE_specs`` is populated with ``comms: local`` and ``nworkers: [num]``.
 
-**"AssertionError: Should not wait for workers when all workers are idle."**
+**"AssertionError: alloc_f did not return any work, although all workers are idle."**
 
-This error occurs when the manager is waiting although no workers are busy, or
-an MPI libEnsemble run was initiated with only one process, resulting in one
-manager but no workers.
+This error occurs when the manager is waiting although no workers are busy.
+Note that a worker can be in a persistent state but is marked as inactive
```
Comment (jmlarson1) [inline docs/FAQ.rst:21]: "inactive" to "idle"

File: docs/executor/overview.rst [hunk 7]
```diff
@@ -113,12 +113,10 @@ The ``app_name`` can be any identfier, while ``full_path`` is the application to
...
-and mechanisms to poll and kill tasks. It will also partition resources amongst
-workers, ensuring that runs utilise different resources (e.g. nodes). The
-``zero_resource_workers`` list option specifies workers that will not need
-resources (e.g. a persistent generator might run on worker 1).
-Furthermore, the MPIExecutor offers resilience via the feature of re-launching
-tasks that fail because of system factors.
+and mechanisms to poll and kill tasks. It also has access to the resource manager,
+which partitions resources amongst workers, ensuring that runs utilise different
+resources (e.g. nodes). Furthermore, the MPIExecutor offers resilience via the
+feature of re-launching tasks that fail because of system factors.
```
Comment (jmlarson1) [inline docs/executor/overview.rst:119]: Should it be "that might have failed"?

File: libensemble/manager.py (hunk 3 above)
Comment (shuds13) [reply libensemble/manager.py:436]: Good catch, this is an old comment. I'm fixing now.

**Comment by jlnav** [comment]
(Cancel out-of-date CI runs so the tests complete faster!)

File: libensemble/alloc_funcs/fast_alloc.py [hunk 8]
```diff
@@ -18,27 +18,39 @@ def give_sim_work_first(W, H, sim_specs, gen_specs, alloc_specs, persis_info):
...
             return_rows = range(len(H)) if gen_in else []
-            gen_work(Work, i, gen_in, return_rows, persis_info.get(i))
+            try:
+                Work[wid] = support.gen_work(wid, gen_in, return_rows, persis_info.get(wid))
+            except InsufficientFreeResources:
+                break
+            gen_count += 1
+            persis_info['total_gen_calls'] += 1
 
+    del support
```
Comment (jmlarson1) [inline libensemble/alloc_funcs/fast_alloc.py:55]: Is this del necessary each call?

File: libensemble/gen_funcs/persistent_uniform_sampling.py [hunk 9]
```diff
@@ -37,3 +37,83 @@ def persistent_uniform(H, persis_info, gen_specs, libE_info):
         H_o['x'] = -1.23
 
     return H_o, persis_info, FINISHED_PERSISTENT_GEN_TAG
+
+
+# SH TODO: Change name of function now
```
Comment (jmlarson1) [inline libensemble/gen_funcs/persistent_uniform_sampling.py:42]: Want to remove?

File: libensemble/gen_funcs/persistent_uniform_sampling.py [hunk 10]
```diff
@@ -37,3 +37,83 @@ def persistent_uniform(H, persis_info, gen_specs, libE_info):
...
+        H_o['x'][i] = x
+        H_o['resource_sets'][i] = 1
+        H_o['priority'] = 1
+
+    # Send batches until manager sends stop tag
+    tag, Work, calc_in = sendrecv_mgr_worker_msg(libE_info['comm'], H_o)
+    while tag not in [STOP_TAG, PERSIS_STOP]:
+        H_o = np.zeros(b, dtype=gen_specs['out'])
+
+        # SH TODO: Re-evaluate - x values - maybe should be clear this is only for a test...
```
Comment (jmlarson1) [inline libensemble/gen_funcs/persistent_uniform_sampling.py:73]: Want to clarify?

File: libensemble/gen_funcs/persistent_uniform_sampling.py [hunk 11]
```diff
@@ -37,3 +37,83 @@ def persistent_uniform(H, persis_info, gen_specs, libE_info):
...
+        # H_o['x'] = len(H)*np.ones(n)
+        H_o['x'] = persis_info['rand_stream'].uniform(lb, ub, (b, n))
+
+        # SH TODO: Should we use persis_info['rand_stream'] for this also?
```
Comment (jmlarson1) [inline libensemble/gen_funcs/persistent_uniform_sampling.py:77]: Yes!

File: libensemble/gen_funcs/sampling.py [hunk 12]
```diff
@@ -29,18 +31,20 @@ def uniform_random_sample_with_different_nodes_and_ranks(H, persis_info, gen_spe
...
+            H_o['resource_sets'][i] = 1
             H_o['priority'] = 1
 
     else:
         H_o = np.zeros(1, dtype=gen_specs['out'])
-        H_o['x'] = len(H)*np.ones(n)
-        H_o['num_nodes'] = persis_info['rand_stream'].randint(1, gen_specs['user']['max_num_nodes']+1)
-        H_o['ranks_per_node'] = persis_info['rand_stream'].randint(1, gen_specs['user']['max_ranks_per_node']+1)
-        H_o['priority'] = 10*H_o['num_nodes']
+        # SH TODO: x values - either random or a basic count
```
Comment (jmlarson1) [inline libensemble/gen_funcs/sampling.py:42]: remove?

File: libensemble/resources/mpi_resources.py [hunk 13]
```diff
@@ -4,158 +4,243 @@
...
+        ranks_per_node = num_procs//num_nodes
+
+    rassert(num_procs == num_nodes*ranks_per_node,
+            "num_procs does not equal num_nodes*ranks_per_node")
+    return num_procs, num_nodes, ranks_per_node
+
+
+def _max_rsets_per_node(worker_resources):
+    """ Return the maximum rsets per node for any node on this worker"""
+    # SH TODO: Need to add tests that go across the uneven nodes.
```
Comment (jmlarson1) [inline libensemble/resources/mpi_resources.py:109]: Are these tests done?

File: libensemble/resources/mpi_resources.py [hunk 14]
```diff
@@ -4,158 +4,243 @@
...
+    local_node_count = wresources.local_node_count
+
+    cores_avail_per_node = \
+        (gresources.logical_cores_avail_per_node if hyperthreads else
+         gresources.physical_cores_avail_per_node)
+
+    rsets_per_node = _max_rsets_per_node(wresources)
+
+    # SH TODO: Evaluate situation when one rset is > 1 node.
+
```
Comment (jmlarson1) [inline libensemble/resources/mpi_resources.py:141]: Want to keep this block of SH TODO notes?

File: libensemble/resources/mpi_resources.py [hunk 15]
```diff
@@ -4,158 +4,243 @@
...
+            num_nodes = local_node_count
+    elif not num_procs and not num_nodes:
+        num_nodes = local_node_count
+
+    # Checks config is consistent and sufficient to express
+    # - does not check actual resources
+    num_procs, num_nodes, ranks_per_node = \
+        task_partition(num_procs, num_nodes, ranks_per_node)
+
+    # SH TODO: Remove
```
Comment (jmlarson1) [inline libensemble/resources/mpi_resources.py:179]: remove?

File: libensemble/resources/resources.py [hunk 16]
```diff
@@ -22,40 +26,91 @@ class ResourcesException(Exception):
...
+    DEFAULT_NODEFILE = 'node_list'
+
+    @classmethod
+    def init_resources(cls, libE_specs):
+        """Initiate resource management"""
+
+        # If disable_resource_manager is True, then Resources.resources will remain None.
+        disable_resource_manager = libE_specs.get('disable_resource_manager', False)
+        if not disable_resource_manager:
+            top_level_dir = os.getcwd()  # SH TODO: Do we want as libE_specs option.
```
Comment (jmlarson1) [inline libensemble/resources/resources.py:62]: I think I'm okay with this not being a libE_spec

File: docs/data_structures/libE_specs.rst (hunk 2 above)
Comment (shuds13) [reply docs/data_structures/libE_specs.rst:97]: Fixing now

File: libensemble/resources/mpi_resources.py (hunk 4 above)
Comment (shuds13) [reply libensemble/resources/mpi_resources.py:39]: yes

File: libensemble/resources/rset_resources.py [hunk 17]
```diff
@@ -0,0 +1,134 @@
...
+            A Resources object containing global nodelist and intranode information
+
+        """
+        self.num_workers = num_workers
+        self.num_workers_2assign2 = RSetResources.get_workers2assign2(self.num_workers, resources)
+        self.total_num_rsets = resources.num_resource_sets or self.num_workers_2assign2
+        self.split_list, self.local_rsets_list = \
+            RSetResources.get_partitioned_nodelist(self.total_num_rsets, resources)
+
+        # SH TODO: Need to update to make uneven distribution of rsets to nodes work as does on develop
```
Comment (jmlarson1) [inline libensemble/resources/rset_resources.py:49]: Has this been done?

File: libensemble/resources/worker_resources.py [hunk 18]
```diff
@@ -0,0 +1,344 @@
...
+            self.rset_team = rset_team
+            self.num_rsets = len(rset_team)
+            self.local_nodelist, self.slots = \
+                WorkerResources.get_local_nodelist(self.workerID, self.rset_team,
+                                                   self.split_list, self.rsets_per_node)
+            self.set_slot_count()
+            self.local_node_count = len(self.local_nodelist)
+
+    # SH TODO: Same count, but I want same list...
+    #          This needs checking... what is slot_count/slots_on_node when uneven
```
Comment (jmlarson1) [inline libensemble/resources/worker_resources.py:274]: Todo?

File: libensemble/alloc_funcs/fast_alloc.py (hunk 8 above)
Comment (shuds13) [reply libensemble/alloc_funcs/fast_alloc.py:55]: I mean instantiated each call

File: libensemble/sim_funcs/six_hump_camel.py [hunk 19]
```diff
@@ -10,9 +11,55 @@
...
+    # Access app output
+    with open('out.txt') as f:
+        H_o['f'] = float(f.readline().strip())  # Read just first line
+
+    calc_status = WORKER_DONE if task.state == 'FINISHED' else 'FAILED'
+    return H_o, persis_info, calc_status
+
+
+# SH TODO: Should we move this to below
+#          Check/update docstring
```
Comment (jmlarson1) [inline libensemble/sim_funcs/six_hump_camel.py:61]: Is the docstring incorrect?

File: libensemble/tests/regression_tests/test_mpi_runners.py [hunk 20]
```diff
@@ -128,16 +136,17 @@
...
-     'mpirun -host node-1,node-2 -np 32 -npernode 16 /path/to/fakeapp.x --testid base2',
-     'mpirun -host node-1,node-2 -np 32 -npernode 16 --xarg 1 /path/to/fakeapp.x --testid base3',
-     'mpirun -host node-1,node-2 -np 128 -npernode 64 --xarg 1 /path/to/fakeapp.x --testid base4',
-     'mpirun -host node-1 -np 16 -npernode 16 --xarg 1 /path/to/fakeapp.x --testid base5',
-     'mpirun -host node-1,node-2 -np 16 -npernode 8 --xarg 1 /path/to/fakeapp.x --testid base6',
-     'mpirun -host node-1 -np 16 --xarg 1 -npernode 16 /path/to/fakeapp.x --testid ompi1',
-     'mpirun -host node-1,node-2 -np 8 --xarg 1 -npernode 4 /path/to/fakeapp.x --testid ompi2',
-     ]
+# openmpi requires machinefiles (-host requires
+# exp_openmpi = \
```
Comment (jmlarson1) [inline libensemble/tests/regression_tests/test_mpi_runners.py:140]: Why comment all of these out?

File: libensemble/tests/regression_tests/test_mpi_runners.py [hunk 21]
```diff
@@ -203,7 +209,8 @@ def run_tests(mpi_runner, runner_name, test_list_exargs, exp_list):
     H, pinfo, flag = libE(sim_specs, gen_specs, exit_criteria, persis_info, libE_specs=libE_specs)
 
 
-for run_set in ['mpich', 'openmpi', 'aprun', 'srun', 'jsrun', 'rename_mpich', 'custom']:
+# for run_set in ['mpich', 'openmpi', 'aprun', 'srun', 'jsrun', 'rename_mpich', 'custom']:
```
Comment (jmlarson1) [inline libensemble/tests/regression_tests/test_mpi_runners.py:212]: Why drop openmpi?

File: libensemble/alloc_funcs/fast_alloc.py (hunk 8 above)
Comment (jmlarson1) [reply libensemble/alloc_funcs/fast_alloc.py:55]: Yes, because it is instantiated each call, is the explicit `del` required?

**Review by jmlarson1 — APPROVED** [review]
I approve. Most of my comments are slight.

File: libensemble/tests/regression_tests/test_mpi_runners.py (hunk 21 above)
Comment (shuds13) [reply libensemble/tests/regression_tests/test_mpi_runners.py:212]: openmpi has to use a machinefile. So we cant test it using runline checks.

File: libensemble/sim_funcs/six_hump_camel.py (hunk 19 above)
Comment (shuds13) [reply libensemble/sim_funcs/six_hump_camel.py]: fixed

File: libensemble/resources/rset_resources.py (hunk 17 above)
Comment (shuds13) [reply libensemble/resources/rset_resources.py]: Yes, but I will add more tests.

File: libensemble/resources/resources.py (hunk 16 above)
Comment (shuds13) [reply libensemble/resources/resources.py]: Yep

File: libensemble/alloc_funcs/fast_alloc.py (hunk 8 above)
Comment (shuds13) [reply libensemble/alloc_funcs/fast_alloc.py:55]: It is not really required. I thought it might be good practice. But lets decide after pull

**Pull request merged**